### DISCUSSION TITLE
Private GitHub Nuget feed authorization issue

### DISCUSSION THREAD

**Discussion by luizbon** [renovatebot/renovate#30027] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

I have a GitHub-hosted Nuget package feed, and the scan can't access it.
I have already tried adding `hostRules` with a custom token, but the scanner has not tried to use it.

From the documentation, I was expecting the scanner would be able to access the packages without any setup, but it can't.

Looking at the logs, it seems the scanner is setting up the GitHub feed token correctly, but further down, the logs show it is trying to access with a blank username and password, resulting in a 401 response.

This is how my `renovate.json` file is:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:recommended"
  ],
  "hostRules": [
    {
      "matchHost": "https://nuget.pkg.github.com/<redacted-org>/index.json",
      "hostType": "nuget",
      "encrypted": {
        "token": "<redacted-token>"
      }
    }
  ]
}
```
The results are the same without `hostRules`.

After the comment from @HonkingGoose, I updated the config to use a username and password, but it is still not working.
The logs are exactly the same as before. Not even the username is used.

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:recommended"
  ],
  "hostRules": [
    {
      "matchHost": "https://nuget.pkg.github.com/<redacted-org>/index.json",
      "hostType": "nuget",
      "username": "<pat-username>",
      "encrypted": {
        "password": "<redacted-token>"
      }
    }
  ]
}
```

It seems my config is completely ignored.

I have a Nuget.config file in the root of the repo with the following content:
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <clear />
    <add key="private-github" value="https://nuget.pkg.github.com/<redacted-org>/index.json" />
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
  </packageSources>
</configuration>
```

I'm running out of ideas about what I am missing.

Thanks.

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: nuget registry failure: can't get RegistrationsBaseUrl
{
  "err": {
    "name": "HTTPError",
    "code": "ERR_NON_2XX_3XX_RESPONSE",
    "timings": {
      "start": 1720071595160,
      "socket": 1720071595160,
      "lookup": 1720071595184,
      "connect": 1720071595279,
      "secureConnect": 1720071595293,
      "upload": 1720071595298,
      "response": 1720071595299,
      "end": 1720071595299,
      "phases": {
        "wait": 0,
        "dns": 24,
        "tcp": 95,
        "tls": 14,
        "request": 5,
        "firstByte": 1,
        "download": 0,
        "total": 139
      }
    },
    "message": "Response code 401 (Unauthorized)",
    "stack": "HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/usr/local/renovate/node_modules/.pnpm/got@11.8.6/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
    "options": {
      "headers": {
        "user-agent": "RenovateBot/37.421.9 (https://github.com/renovatebot/renovate)",
        "accept": "application/json",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://nuget.pkg.github.com/<redacted-org>/index.json",
      "hostType": "nuget",
      "username": "",
      "password": "",
      "method": "GET",
      "http2": false
    },
    "response": {
      "statusCode": 401,
      "statusMessage": "Unauthorized",
      "body": "",
      "headers": {
        "access-control-allow-methods": "GET, OPTIONS",
        "access-control-allow-origin": "*",
        "content-security-policy": "default-src 'none';",
        "server": "GitHub Registry",
        "strict-transport-security": "max-age=31536000;",
        "www-authenticate": "Basic realm=\"GitHub Package Registry\"",
        "x-content-type-options": "nosniff",
        "x-frame-options": "DENY",
        "x-nuget-warning": "No Authorization header detected",
        "x-xss-protection": "1; mode=block",
        "date": "Thu, 04 Jul 2024 05:39:55 GMT",
        "content-length": "0",
        "x-github-request-id": "8822:19023:223CC1:28FA31:668635AB",
        "connection": "close"
      },
      "httpVersion": "1.1",
      "retryCount": 0
    }
  }
  "url": "https://nuget.pkg.github.com/<redacted-org>/index.json"
}

```

</details>

**Comment by HonkingGoose** [comment]
Maybe these Renovate docs quotes can help you?

> ## When does Renovate need credentials?
> 
> By default, the only credentials Renovate has are those for the "platform", i.e. GitHub, GitLab, etc.
> If the token used has sufficient permissions, this will enable Renovate to lookup dependencies located in alternative repositories on the same host or any hosted on any embedded package registry on the same host.
> 
> It's also quite common to need to look up packages on other protected hosts, including npmjs, Docker Hub, or private registries like Nexus or Artifactory.
> Any time you need Renovate to access such registries with credentials then you will need to provision them as part of your config.
> 
> There are four times in Renovate's behavior when it may need credentials:
> 
> - Resolving private config presets
> - Looking up dependency versions
> - Looking up changelogs
> - Passing to package managers when updating lock files or checksums
>
> NOTE
> If you self-host Renovate, and have a self-hosted registry which _doesn't_ require authentication to access, then such modules/packages are not considered "private" to Renovate.

> ### nuget (package manager credentials for artifact updating)
> 
> For each known NuGet registry, Renovate searches for `hostRules` with `hostType=nuget` and matching host.
> For those found, a command like the following is run: `dotnet nuget add source ${registryInfo.feedUrl} --configfile ${nugetConfigFile} --username ${username} --password ${password} --store-password-in-clear-text`.
> 
> ```js
> hostRules: [
>   {
>     matchHost: 'https://pkgs.dev.azure.com/<org>/',
>     hostType: 'nuget',
>     username: 'user', // doesn't matter for azure
>     password: '<PAT>',
>   },
> ];
> ```

## References

The quotes above come from these Renovate docs sections:

- [Private package support, when does Renovate need credentials?](https://docs.renovatebot.com/getting-started/private-packages/#when-does-renovate-need-credentials)
- [Private packages, package manager credentials for artifact updating, nuget](https://docs.renovatebot.com/getting-started/private-packages/#nuget)

**Reply by luizbon** [reply]
Good callout, @HonkingGoose. I was setting it up as a token instead of a username and password.
Unfortunately, even after changing to a username and password, the setup seems to be ignored, and the request to the GitHub NuGet feed is still sending an empty username and password.

I'll update the original post with the new config.

**Reply by viceice** [reply]
remove the `index.json` from match host. otherwise only that single url is authorized
